### PR TITLE
[evm] EIP-1153 enable transient storage feature 

### LINKED ISSUE #4177: [evm] enable transient storage feature
### Description of new feature request
EIP-1153 transient storage is soon to be activated in Cancun upgrade. We will enable it in our EVM in upcoming (or the next afterwards) hard-fork
1. #4151 has added the necessary data structure
2. continue to enable it in EVM. Please refer to https://github.com/ethereum/go-ethereum/pull/26003 for correct implementation

### Is this a hard-fork (and height to be activated)
- [x] Yes
- Name of activation height:

### Implementation plan
- [] Phase 1 description (PR number)
- [] Phase 2 description (PR number)
- [] Phase 3 description (PR number)

### Affected modules
- [] action
- [] actpool
- [] api
- [x] blockchain
- [] blockindex
- [] blocksync
- [] chainservice
- [] config
- [] consensus
- [] db
- [] dispatcher
- [] ioctl
- [] p2p
- [] server
- [] state
- [] tools

### PR DESCRIPTION
**Pull request by millken**
# Description
Implemented [EIP-1153](https://eips.ethereum.org/EIPS/eip-1153)

Fixes #4177

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/execution/evm/evm_test.go [hunk 1]
```diff
@@ -259,7 +259,7 @@ func TestConstantinople(t *testing.T) {
 		},
 		{
 			"io1pcg2ja9krrhujpazswgz77ss46xgt88afqlk6y",
-			1261440000, // = 200*365*24*3600/5, around 200 years later
+			39275560,
```
Comment (millken) [inline action/protocol/execution/evm/evm_test.go:262]: it must less genesis default UpernavikBlockHeight

File: action/protocol/execution/testdata-cancun/transientstorage.sol [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    function tstore(uint location, uint value) public {
+        assembly {
+            tstore(location, value)
+        }
+    }
+
+    function tload(uint location) public view returns (uint value) {
+        assembly {
+            value := tload(location)
+        }
```
Comment (millken) [inline action/protocol/execution/testdata-cancun/transientstorage.sol:28]: Transient storage is accessible to smart contracts via 2 new opcodes, `TLOAD` and `TSTORE`

refer: https://eips.ethereum.org/EIPS/eip-1153

File: action/protocol/execution/evm/evm_test.go (hunk 1 above)
Comment (dustinxie) [reply action/protocol/execution/evm/evm_test.go:262]: check past PR, you need to break it to 2 parts
```
after Tsunami - Upernavik {
}
after Upernavik {
}
```

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 3]
```diff
@@ -478,13 +483,22 @@ func (stateDB *StateDBAdapter) HasSelfDestructed(evmAddr common.Address) bool {
...
+	prev := stateDB.GetTransientState(addr, key)
+	if prev == value {
+		return
+	}
+	stateDB.setTransientState(addr, key, value)
+}
+
+// setTransientState is a lower level setter for transient storage. It
+// is called during a revert to prevent modifications to the journal.
+func (stateDB *StateDBAdapter) setTransientState(addr common.Address, key, value common.Hash) {
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:495]: no need to add this new func? it's only used at L490 above

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 4]
```diff
@@ -478,13 +483,22 @@ func (stateDB *StateDBAdapter) HasSelfDestructed(evmAddr common.Address) bool {
 
 // SetTransientState sets transient storage for a given account
 func (stateDB *StateDBAdapter) SetTransientState(addr common.Address, key, value common.Hash) {
-	log.S().Panic("SetTransientState not implemented")
+	prev := stateDB.GetTransientState(addr, key)
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:486]: interface method should not call another interface method
use `stateDB.transientStorage.Get(addr, key)` instead

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 5]
```diff
@@ -530,6 +544,8 @@ func (stateDB *StateDBAdapter) Prepare(rules params.Rules, sender, coinbase comm
 	if !rules.IsBerlin {
 		return
 	}
+	// Clear out any leftover from previous executions
+	stateDB.accessList = newAccessList()
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:548]: this is not necessary? we are creating a new `StateDBAdpator` for each tx

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 6]
```diff
@@ -547,6 +563,8 @@ func (stateDB *StateDBAdapter) Prepare(rules params.Rules, sender, coinbase comm
 	if rules.IsShanghai { // EIP-3651: warm coinbase
 		stateDB.AddAddressToAccessList(coinbase)
 	}
+	// Reset transient storage at the beginning of transaction execution
+	stateDB.transientStorage = newTransientStorage()
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:567]: same thing, not necessary

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 7]
```diff
@@ -592,6 +610,23 @@ func (stateDB *StateDBAdapter) RevertToSnapshot(snapshot int) {
 		stateDB.logError(err)
 		return
 	}
+	// restore transientStorage
+	tss, ok := stateDB.transientStorageSnapshot[snapshot]
+	if !ok {
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:615]: it is checked below, no need to do another check

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 8]
```diff
@@ -592,6 +610,23 @@ func (stateDB *StateDBAdapter) RevertToSnapshot(snapshot int) {
...
+	{
+		delete(stateDB.transientStorageSnapshot, snapshot)
+		for i := snapshot + 1; ; i++ {
+			if _, ok := stateDB.transientStorageSnapshot[i]; ok {
+				delete(stateDB.transientStorageSnapshot, i)
+			} else {
+				break
+			}
+		}
+	}
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:629]: move L619-629 to after `// restore access list`

File: action/protocol/execution/evm/evmstatedbadapter_test.go [hunk 9]
```diff
@@ -934,3 +934,45 @@ func TestSortMap(t *testing.T) {
...
+	state.SetTransientState(addr, key, value)
+	if exp, got := 0, sn; exp != got {
+		t.Fatalf("journal length mismatch: have %d, want %d", got, exp)
+	}
+	// the retrieved value should equal what was set
+	if got := state.GetTransientState(addr, key); got != value {
+		t.Fatalf("transient storage mismatch: have %x, want %x", got, value)
+	}
+
+	// revert the transient state being set and then check that the
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter_test.go:966]: for the revert part, need to add test case/data for transient storage into existing `TestSnapshotRevertAndCommit`

File: action/protocol/execution/testdata-cancun/transientstorage.sol [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    }
+}
```
Comment (dustinxie) [inline action/protocol/execution/testdata-cancun/transientstorage.sol:30]: add a new line at the end

File: action/protocol/execution/evm/evmstatedbadapter_test.go (hunk 9 above)
Comment (dustinxie) [reply action/protocol/execution/evm/evmstatedbadapter_test.go:966]: for this test, can use the `addr, key, value` test data in `TestTransientStorage` to verify it works correctly

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 11]
```diff
@@ -750,6 +785,7 @@ func (stateDB *StateDBAdapter) Snapshot() int {
 	stateDB.preimageSnapshot[sn] = p
 	// save a copy of access list
 	stateDB.accessListSnapshot[sn] = stateDB.accessList.Copy()
+	stateDB.transientStorageSnapshot[sn] = stateDB.transientStorage.Copy()
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:788]: add comment

File: action/protocol/execution/evm/evmstatedbadapter.go (hunk 5 above)
Comment (millken) [reply action/protocol/execution/evm/evmstatedbadapter.go:548]: for safety, it's best to add

File: action/protocol/execution/evm/evmstatedbadapter.go (hunk 6 above)
Comment (millken) [reply action/protocol/execution/evm/evmstatedbadapter.go:567]: it is best to add

File: action/protocol/execution/evm/evmstatedbadapter.go (hunk 5 above)
Comment (envestcc) [reply action/protocol/execution/evm/evmstatedbadapter.go:542]: What is the relationship with eip-1153?

File: action/protocol/execution/evm/evm_test.go [hunk 12]
```diff
@@ -356,7 +356,8 @@ func TestConstantinople(t *testing.T) {
 		require.Equal(isSumatra, chainRules.IsShanghai)
 
 		// Cancun, Prague not yet enabled
-		require.False(evmChainConfig.IsCancun(big.NewInt(int64(e.height)), evm.Context.Time))
+		isUpernavik := g.IsUpernavik(e.height)
```
Comment (envestcc) [inline action/protocol/execution/evm/evm_test.go:359]: comments need updates

File: action/protocol/execution/evm/evmstatedbadapter_test.go [hunk 13]
```diff
@@ -949,30 +977,46 @@ func TestStateDBTransientStorage(t *testing.T) {
...
+		{addr1, k2, v1},
 	}
+	for _, test := range tests {
+		addr := test.addr
+		key := test.key
+		value := test.val
+		sn := state.Snapshot()
+		state.SetTransientState(addr, key, value)
+		// the retrieved value should equal what was set
+		if got := state.GetTransientState(addr, key); got != value {
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter_test.go:1004]: let's be consistent with testing code style, use `require.Equal(value, got)`

File: action/protocol/execution/evm/evmstatedbadapter_test.go [hunk 14]
```diff
@@ -949,30 +977,46 @@ func TestStateDBTransientStorage(t *testing.T) {
...
-	// value is now the empty hash
-	state.RevertToSnapshot(sn)
-	if got, exp := state.GetTransientState(addr, key), (common.Hash{}); exp != got {
-		t.Fatalf("transient storage mismatch: have %x, want %x", got, exp)
-	}
+		// revert the transient state being set and then check that the
+		// value is now the empty hash
+		state.RevertToSnapshot(sn)
+		if got, exp := state.GetTransientState(addr, key), (common.Hash{}); exp != got {
+			t.Fatalf("transient storage mismatch: have %x, want %x", got, exp)
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter_test.go:1012]: same here, use current testing code style

File: action/protocol/execution/evm/evmstatedbadapter_test.go [hunk 15]
```diff
@@ -949,30 +977,46 @@ func TestStateDBTransientStorage(t *testing.T) {
...
+		}
 
-	// reset transient state
-	state.SetTransientState(addr, key, value)
-	if got := state.GetTransientState(addr, key); got != value {
-		t.Fatalf("transient storage mismatch: have %x, want %x", got, value)
+		// reset transient state
+		state.SetTransientState(addr, key, value)
+		if got := state.GetTransientState(addr, key); got != value {
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter_test.go:1017]: same here

File: action/protocol/execution/evm/evmstatedbadapter_test.go [hunk 16]
```diff
@@ -677,6 +704,7 @@ func TestSnapshotRevertAndCommit(t *testing.T) {
 			// refund fix and accessList are introduced after fixSnapshot
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter_test.go:704]: update comment, add transient storage

File: action/protocol/execution/evm/evmstatedbadapter.go (hunk 5 above)
Comment (CoderZhi) [reply action/protocol/execution/evm/evmstatedbadapter.go:542]: > for safety, it's best to add

for what kind of safety? We need to be clear what we want to achieve.

Comment (millken) [reply action/protocol/execution/evm/evmstatedbadapter.go:542]: the code has updated in
https://github.com/ethereum/go-ethereum/pull/26003/files#diff-c3757dc9e9d868f63bc84a0cc67159c1d5c22cc5d8c9468757098f0492e0658cR1073-R1074

i think we missed it

File: action/protocol/execution/evm/evm_test.go (hunk 1 above)
Comment (dustinxie) [reply action/protocol/execution/evm/evm_test.go]: still not fixed?

File: action/protocol/execution/testdata-cancun/transientstorage.json [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+        "rawGasLimit": 1000000,
+        "rawGasPrice": "0",
+        "rawAccessList": [],
+        "rawExpectedGasConsumed": 11068,
+        "expectedStatus": 1,
+        "readOnly": true,
+        "rawReturnValue": "0000000000000000000000000000000000000000000000000000000000000003",
+        "comment": "call getAddress"
+    }]
+}
```
Comment (dustinxie) [inline action/protocol/execution/testdata-cancun/transientstorage.json:49]: add newline

File: action/protocol/execution/evm/evmstatedbadapter_test.go [hunk 18]
```diff
@@ -701,7 +701,7 @@ func TestSnapshotRevertAndCommit(t *testing.T) {
 			require.Equal(3, len(stateDB.contractSnapshot))
 			require.Equal(3, len(stateDB.selfDestructedSnapshot))
 			require.Equal(3, len(stateDB.preimageSnapshot))
-			// refund fix and accessList、transient storage are introduced after fixSnapshot
+			// refund fix and accessList,transient storage are introduced after fixSnapshot
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter_test.go:704]: refund fix, accessList, and transient

File: action/protocol/execution/evm/evm_test.go (hunk 1 above)
Comment (millken) [reply action/protocol/execution/evm/evm_test.go]: fix now

File: action/protocol/execution/evm/evm_test.go [hunk 19]
```diff
@@ -257,6 +257,16 @@ func TestConstantinople(t *testing.T) {
 			action.EmptyAddress,
 			29275561,
 		},
+		// after Tsunami - Upernavik
+		{
+			action.EmptyAddress,
```
Comment (dustinxie) [inline action/protocol/execution/evm/evm_test.go:262]: should be "io1pcg2ja9krrhujpazswgz77ss46xgt88afqlk6y"

File: action/protocol/execution/evm/evm_test.go [hunk 20]
```diff
@@ -257,6 +257,16 @@ func TestConstantinople(t *testing.T) {
 			action.EmptyAddress,
 			29275561,
 		},
+		// after Tsunami - Upernavik
```
Comment (dustinxie) [inline action/protocol/execution/evm/evm_test.go:260]: and this comment line should be at L255

File: action/protocol/execution/evm/evmstatedbadapter.go (hunk 6 above)
Comment (dustinxie) [reply action/protocol/execution/evm/evmstatedbadapter.go:559]: as discused, let's remove this, it is done in `NewStateDBAdapter`

**Review by dustinxie — APPROVED** [review]

**Review by envestcc — APPROVED** [review]

**Pull request merged**